### PR TITLE
[stdlib] Speed up bridged Dictionary instances

### PR DESCRIPTION
**Pull request by lorentey**
`Dictionary`’s native storage classes and `_SwiftDeferredNSDictionary` override `-[NSDictionary getObjects:andKeys:]` instead of its safer replacement, `-[NSDictionary getObjects:andKeys:count:]`.

Overriding the correct method will considerably speed up some Cocoa operations on bridged dictionaries.

rdar://problem/39285882

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
I don't think we have benchmarks for bridged dictionaries yet, but just in case:

@swift-ci please benchmark

File: stdlib/public/core/Dictionary.swift [hunk 1]
```diff
@@ -2130,28 +2133,33 @@ final internal class _HashableTypedNativeDictionaryStorage<Key: Hashable, Value>
...
+    andKeys keys: UnsafeMutablePointer<AnyObject>?,
+    count: Int) {
     // The user is expected to provide a storage of the correct size
     if let unmanagedKeys = _UnmanagedAnyObjectArray(keys) {
       if let unmanagedObjects = _UnmanagedAnyObjectArray(objects) {
         // keys nonnull, objects nonnull
         for (offset: i, element: (key: key, value: val)) in full.enumerated() {
           unmanagedObjects[i] = _bridgeAnythingToObjectiveC(val)
           unmanagedKeys[i] = _bridgeAnythingToObjectiveC(key)
+          guard i < count else { break }
```
Comment (jtbandes) [inline stdlib/public/core/Dictionary.swift]: Sorry for the drive-by comment, but shouldn’t this be checked _before_ assigning at position `i`? (Same applies below)

**Comment by lorentey** [comment]
I added new microbenchmarks to help us track performance. Copying elements in bulk out of an `_SwiftDeferredNSDictionary` is now ~5 times faster:

<details open>
  <summary>Improvement (1)</summary>
  
TEST                              | OLD | NEW | DELTA  | SPEEDUP   
---                               | --- | --- | ---    | ---       
DictionaryBridgeToObjC_BulkAccess | 699 | 143 | -79.5% | **4.89x** 

</details>

<details open>
  <summary>No Changes (2)</summary>
  
TEST                              | OLD | NEW | DELTA  | SPEEDUP   
---                               | --- | --- | ---    | ---       
DictionaryBridgeToObjC_Access     | 774 | 760 | -1.8%  | 1.02x (?) 
DictionaryBridgeToObjC_Bridge     | 18  | 18  | +0.0%  | 1.00x     

</details>

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
@swift-ci test

File: stdlib/public/core/Dictionary.swift [hunk 2]
```diff
@@ -2961,6 +2975,7 @@ final internal class _SwiftDeferredNSDictionary<Key: Hashable, Value>
         // keys null, objects nonnull
         for position in 0..<bucketCount {
           if bridgedBuffer.isInitializedEntry(at: position) {
+            guard i < count else { break }
```
Comment (michael-lehew) [inline stdlib/public/core/Dictionary.swift]: you could eke out a one less loop iteration by checking this after you increment i -- and adding an early out above for count == 0

**Comment by moiseev** [comment]
:shipit:

File: stdlib/public/core/Dictionary.swift (hunk 2 above)
Comment (lorentey) [reply stdlib/public/core/Dictionary.swift]: Good point! It can often be more than one iteration, too, since the loop skips over empty buckets.

**Review by lancep — APPROVED** [review]
LGTM! 🍔

**Comment by lorentey** [comment]
@swift-ci test and merge
